aspnetcore/blazor/host-and-deploy/configure-trimmer.md

```diff
@@ -42,6 +40,141 @@ The default trimmer granularity for Blazor apps is `partial`. To trim all assemb
 
 For more information, see [Trimming options (.NET documentation)](/dotnet/core/deploying/trimming/trimming-options#trimming-granularity).
 
+## Failure to preserve types used by a published app
+
+Trimming may have detrimental effects for a published app leading to runtime errors. In apps that use [reflection](/dotnet/csharp/advanced-topics/reflection-and-attributes/), the IL Trimmer often can't determine the required types for runtime reflection and trims them away or trims away parameter names from methods. This can happen with complex framework types used for JS interop, JSON serialization/deserialization, and other operations.
+
+The IL Trimmer is also unable to react to an app's dynamic behavior at runtime. To ensure the trimmed app works correctly once deployed, test published output frequently while developing.
+
+Consider the following client-side component in a Blazor Web App (ASP.NET Core 8.0 or later) that deserializes a <xref:System.Collections.Generic.KeyValuePair> collection (`List<KeyValuePair<string, string>>`):
+
+```razor
+@rendermode @(new InteractiveWebAssemblyRenderMode(false))
+@using System.Diagnostics.CodeAnalysis
+@using System.Text.Json
+
+<dl>
+    @foreach (var item in @items)
+    {
+        <dt>@item.Key</dt>
+        <dd>@item.Value</dd>
+    }
+</dl>
+
+@code {
+    private List<KeyValuePair<string, string>> items = [];
+
+    [StringSyntax(StringSyntaxAttribute.Json)]
+    private const string data =
+        """[{"key":"key 1","value":"value 1"},{"key":"key 2","value":"value 2"}]""";
+
+    protected override void OnInitialized()
+    {
+        JsonSerializerOptions options = new() { PropertyNameCaseInsensitive = true };
+
+        items = JsonSerializer
+            .Deserialize<List<KeyValuePair<string, string>>>(data, options)!;
+    }
+}
+```
+
+The preceding component executes normally when the app is run locally and produces the following rendered definition list (`<dl>`):
+
+> **:::no-loc text="key 1":::**  
+> :::no-loc text="value 1":::  
+> **:::no-loc text="key 2":::**  
+> :::no-loc text="value 2":::
+
+When the app is published, <xref:System.Collections.Generic.KeyValuePair> is trimmed from the app, even in spite of setting the [`<PublishTrimmed>` property](#configuration) to `false` in the project file. Accessing the component throws the following exception:
+
+> :::no-loc text="Unhandled exception rendering component: ConstructorContainsNullParameterNames, System.Collections.Generic.KeyValuePair`2[System.String,System.String]":::
+
+To address lost types, consider the following approaches.
```

Comment:
~~approaches~~ -> solution

In singular until we un-comment the rest

Reply:
There are two ... the dynamic dep attribute and the custom type.

Reply:
The custom type is currently written out with the example. The commented text for it will replace that when the descriptor approach comes back. Either way, this sentence will always indicate more than one approach.

Reply:
ok